src/System.Reflection.Metadata.ILDasm/src/Decoder/ILDecoder.cs

```diff
@@ -0,0 +1,582 @@
+        {
+            if (IsStandaloneSignature(intOperand))
+            {
+                var handle = MetadataTokens.StandaloneSignatureHandle(intOperand);
+                var standaloneSignature = mdReader.GetStandaloneSignature(handle);
+                var signature = SignatureDecoder.DecodeMethodSignature(standaloneSignature.Signature, provider);
+                return string.Format("{0}{1}", GetMethodReturnType(signature), provider.GetParameterList(signature));
+            }
+            throw new ArgumentException("Get signature invalid token");
+        }
+
+        private static string GetVariableName(OpCode opCode,int token, ILMethodDefinition _methodDefinition)
+        {
+            if (HasArgument(opCode))
+            {
+                if (_methodDefinition.Signature.Header.IsInstance)
+                {
+                    token--; //the first parameter is "this".
+                }
+                return _methodDefinition.GetParameter(token).Name;
+            }
+            return _methodDefinition.GetLocal(token).Name;
+            
+        }
+
+        private static string GetInlineTokenType(MetadataReader mdReader, int intOperand, ILTypeProvider provider)
+        {
+            if(IsMethodDefinition(intOperand) || IsMethodSpecification(intOperand) || IsMemberReference(intOperand))
+            {
+                return "method " + SolveMethodName(mdReader, intOperand, provider);
+            }
+            if (IsFieldDefinition(intOperand))
+            {
+                return "field " + GetFieldInformation(mdReader, intOperand, provider);
+            }
+            return GetTypeInformation(mdReader, intOperand, provider);
+        }
+
+        private static string GetTypeInformation(MetadataReader mdReader, int intOperand, ILTypeProvider provider)
+        {
+            if(IsTypeReference(intOperand))
+            {
+                var refHandle = MetadataTokens.TypeReferenceHandle(intOperand);
+                return SignatureDecoder.DecodeType(refHandle, provider).ToString();
+            }
+            if (IsTypeSpecification(intOperand))
+            {
+                var typeHandle = MetadataTokens.TypeSpecificationHandle(intOperand);
+                return SignatureDecoder.DecodeType(typeHandle, provider).ToString();
+            }
+            var defHandle = MetadataTokens.TypeDefinitionHandle(intOperand);
+            return SignatureDecoder.DecodeType(defHandle, provider).ToString();
+        }
+
+        private static ILInstruction CreateSwitchInstruction(ref BlobReader ilReader, int expectedSize, int ilOffset, OpCode opCode)
+        {
+            var caseNumber = ilReader.ReadUInt32();
+            int[] jumps = new int[caseNumber];
+            for(int i = 0; i < caseNumber; i++)
+            {
+                jumps[i] = ilReader.ReadInt32();
+            }
+            int size = 4 + expectedSize;
```

Comment:
It'd be good to put values like the ```4```s here into named constants.  As a reader not deeply familiar with this space, I'm not sure what these values mean and why they're correct.